src/meta/src/barrier/recovery.rs

```diff
@@ -386,6 +397,82 @@ impl GlobalBarrierManager {
         Ok(true)
     }
 
+    async fn scale_actors(&self, info: &BarrierActorInfo) -> MetaResult<bool> {
+        debug!("start scaling-in offline actors.");
+
+        let expired_workers: HashSet<WorkerId> = info
+            .actor_map
+            .iter()
+            .filter(|(&worker, actors)| !actors.is_empty() && !info.node_map.contains_key(&worker))
+            .map(|(&worker, _)| worker)
+            .collect();
+
+        if expired_workers.is_empty() {
+            debug!("no expired workers, skipping.");
+            return Ok(false);
+        }
+
+        let all_worker_parallel_units = self.fragment_manager.all_worker_parallel_units().await;
+
+        let expired_worker_parallel_units: HashMap<_, _> = all_worker_parallel_units
+            .into_iter()
+            .filter(|(worker, _)| expired_workers.contains(worker))
+            .collect();
+
+        let fragment_worker_changes = {
+            let guard = self.fragment_manager.get_fragment_read_guard().await;
+            let mut policy = HashMap::new();
+            for table_fragments in guard.table_fragments().values() {
+                for fragment_id in table_fragments.fragment_ids() {
+                    policy.insert(
+                        fragment_id,
+                        PbWorkerChanges {
+                            exclude_worker_ids: expired_workers.iter().cloned().collect(),
+                            ..Default::default()
+                        },
+                    );
+                }
+            }
+            policy
+        };
+
+        let plan = self
+            .scale_controller
+            .generate_stable_resize_plan(
+                StableResizePolicy {
+                    fragment_worker_changes,
+                },
+                Some(expired_worker_parallel_units),
+            )
+            .await?;
+
+        let (reschedule_fragment, applied_reschedules) = self
+            .scale_controller
+            .prepare_reschedule_command(
+                plan,
+                RescheduleOptions {
+                    resolve_no_shuffle_upstream: true,
+                },
+            )
+            .await?;
+
+        match self
+            .scale_controller
+            .post_apply_reschedule(&reschedule_fragment)
+            .await
+        {
+            Ok(_) => {}
+            Err(_e) => {
```

Comment:
Better to add some debug logs here. It could be request too large error, the logs can be helpful to identify the problem.